src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,610 @@
+    // we'll need to disable this mode at compile-time by removing it or making
+    // it a compile-time feature.
+    Native,
+    /// This mode assumes that libc + Linux's native thread local storage is
+    /// being used in a conventional way, but doesn't actually store any data in
+    /// it. Instead, it uses the base address of the current thread's native
+    /// thread local storage as a "fast thread ID" to index into statically
+    /// allocated storage.
+    ///
+    /// If the current thread doesn't appear to have native thread local storage,
+    /// falls back to `Gettid`.
+    ///
+    /// This mode is close to the speed of `Native`, but avoids some potentially
+    /// interactions with libc, and doesn't require linking with libc.
+    ///
+    /// SAFETY: Requires that each thread using this thread local storage
+    /// calls [`unregister_current_thread`] before exiting.
+    NativeTlsId,
+    /// Use each thread's native identifier to index into statically allocated
+    /// thread local storage. This mode doesn't depend on any libc or any
+    /// assumptions about libc, but is substantially slower than the other modes
+    /// since a `gettid` syscall must be executed on every access to a
+    /// thread-local variable.
+    #[allow(unused)]
+    Gettid,
+}
+
+pub mod global_preferred_mode {
+    use super::*;
+
+    static PREFERRED_MODE: AtomicI8 = AtomicI8::new(Mode::Native as i8);
+
+    /// # Safety
+    ///
+    /// There must not exist any live threads that have accessed Tls variables
+    /// and not yet called [`unregister_current_thread`].
+    ///
+    /// See [`Mode`] for addititional requirements for each mode.
+    pub unsafe fn set(mode: Mode) {
+        PREFERRED_MODE.store(mode.into(), atomic::Ordering::Relaxed);
+    }
+
+    pub fn get() -> Mode {
+        PREFERRED_MODE
+            .load(atomic::Ordering::Relaxed)
+            .try_into()
+            .unwrap()
+    }
+}
+
+// This needs to be big enough to store all thread-local variables for a single
+// thread. We fail at runtime if this limit is exceeded.
+pub const BYTES_PER_THREAD: usize = 1024;
+
+// Max threads for our slow TLS fallback mechanism.  We support recycling
+// storage of exited threads, so this is the max *concurrent* threads per
+// process.
+const TLS_FALLBACK_MAX_THREADS: usize = 1000;
+
+// This is `pub` for export to C, where we can allocate
+// native thread local storage.
+#[repr(C, align(16))]
+pub struct ShimThreadLocalStorage {
+    _bytes: [u8; BYTES_PER_THREAD],
+}
+
+/// An opaque, per-thread identifier. These are only guaranteed to be unique for
+/// *live* threads; in particular `FastThreadId::Fs` of a live thread can have
+/// the same value as a previously seen dead thread. See
+/// [`unregister_current_thread`].
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Hash)]
+pub(crate) enum FastThreadId {
+    None,
+    Fs(usize),
+    NativeTid(Pid),
+}
```

Comment:
We're not permitted to implement `From<Option<FastThreadId>> for usize`, but on further consideration I think it is better to move this outside and just turn that conversion into a method instead of using the `From` trait.